code/datums/progress_bar.dm

```diff
@@ -0,0 +1,213 @@
+	var/goal = 100
+	/// The icon file to use for the visual element.
+	var/icon
+	/// The total height in pixels of the theme's progress bar.
+	var/height
+	/// The width in pixels of the theme's fill sprite.
+	var/fill_width
+	/// The total time that determines the speed for appear animation.
+	var/animation_time_appear = 0.5 SECOND
+	/// The total time that determines the speed for fade animation.
+	var/animation_time_fade = 0.5 SECOND
+	/// The total time that determines the speed for shift down animation.
+	var/animation_time_shift = 0.5 SECOND
+
+	// Variables
+	/// The target to display the progress bar on.
+	var/atom/target
+	/// The main visual element holding everything up.
+	var/image/holder
+	/// The backdrop visual element. Contains the fill and mask as overlays.
+	var/image/backdrop
+	/// The fill visual element.
+	var/image/fill
+	/// The mob currently owning the progress bar, used to calculate the bar's vertical position.
+	var/mob/user
+	/// The client currently owning the progress bar for display.
+	var/client/client
+	/// The bar's index number within the owning mob's active progress bar list.
+	var/list_index
+	/// Whether the bar is stopping.
+	var/stopping = FALSE
+	/// Stored progress number for the goal fail effect.
+	var/last_progress = 0
+
+/datum/progress_bar/New(mob/param_user, param_goal, atom/param_target)
+	. = ..()
+	if(!istype(param_target))
+		throw EXCEPTION("datum/progress_bar: Invalid param given - param_target")
+
+	target = param_target
+	user = param_user
+	goal = param_goal
+	init_images(target)
+
+	if(user)
+		client = user.client
+		client?.images |= holder
+
+		LAZYINITLIST(user.progress_bars)
+		LAZYINITLIST(user.progress_bars[target])
+		var/list/bars = user.progress_bars[target]
+		bars += src
+		list_index = length(bars)
+		animate(holder, pixel_z = world.icon_size + (height * (list_index - 1)), alpha = 255, time = animation_time_appear, easing = SINE_EASING | EASE_OUT, flags = ANIMATION_PARALLEL)
+	else
+		animate(holder, alpha = 255, time = animation_time_appear, easing = SINE_EASING | EASE_OUT, flags = ANIMATION_PARALLEL)
+
+/datum/progress_bar/Destroy()
+	stop()
+	client?.images -= holder
+	QDEL_NULL(holder)
+	QDEL_NULL(backdrop)
+	QDEL_NULL(fill)
+	target = null
+	user = null
+	client = null
+	return ..()
+
+/**
+  * Stops the progress bar and removes it after a bit.
+  */
+/datum/progress_bar/proc/stop()
+	if(stopping)
+		return
+
+	if(user?.progress_bars[target])
+		var/list/bars = user.progress_bars[target]
+		var/list_index = bars.Find(src)
+		for(var/i = list_index, i <= length(bars), i++)
+			var/datum/progress_bar/P = bars[i]
+			P.shift()
+
+		bars -= src
+		if(!length(bars))
+			LAZYREMOVE(user.progress_bars, target)
+
+	if(!QDELING(src))
+		stopping = TRUE
+		update(last_progress)
+		animate(holder, alpha = 0, time = animation_time_fade, flags = ANIMATION_PARALLEL)
+		QDEL_IN(src, animation_time_fade)
```

Comment:
Think after's previous comment about sleeping rather than using qdel_in might still apply?

Reply:
I don't know a benefits of sleep to make the corresponding change.

Reply:
I think `sleep` is simpler and naturally prefer it (note that you'd need to make this proc waitfor-FALSE for that to work). I don't have strong reasoning beyond that, and would not consider it blocking.